src/array/array.cc

```diff
@@ -652,6 +653,80 @@ std::pair<COOMatrix, IdArray> COOCoalesce(COOMatrix coo) {
   return ret;
 }
 
+///////////////////////// Graph Traverse  routines //////////////////////////
+Frontiers BFSNodesFrontiers(const CSRMatrix& csr, IdArray source) {
+  Frontiers ret;
+  CHECK_EQ(csr.indptr->ctx.device_type, source->ctx.device_type) <<
+    "Graph and source should in the same device context";
+  CHECK_EQ(csr.indices->dtype, source->dtype) <<
+    "Graph and source should in the same dtype";
```

Comment:
There is a `CHECK_SAME_DTYPE` and a `CHECK_SAME_CONTEXT` macros in my PR. It could help with more consistent error messages.

Reply:
`CHECK_EQ(csr.num_rows, csr.num_cols) << "Graph traversal can only work on square-shaped CSR."`.

Reply:
> There is a `CHECK_SAME_DTYPE` and a `CHECK_SAME_CONTEXT` macros in my PR. It could help with more consistent error messages.

Let's merge your PR first.